File: otel-tests/nodejs/tests/otel.test.js
/**
 * Copyright 2023 Greptime Team
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

const { metrics, trace, context, SpanKind, SpanStatusCode } = require('@opentelemetry/api');
const { logs, SeverityNumber } = require('@opentelemetry/api-logs');
const { Resource } = require('@opentelemetry/resources');
const {
  SEMRESATTRS_SERVICE_NAME,
  SEMRESATTRS_SERVICE_VERSION,
} = require('@opentelemetry/semantic-conventions');
const {
  MeterProvider,
  PeriodicExportingMetricReader,
} = require('@opentelemetry/sdk-metrics');
const {
  BasicTracerProvider,
  BatchSpanProcessor,
} = require('@opentelemetry/sdk-trace-base');
const {
  LoggerProvider,
  BatchLogRecordProcessor,
} = require('@opentelemetry/sdk-logs');
const {
  OTLPMetricExporter,
} = require('@opentelemetry/exporter-metrics-otlp-proto');
const {
  OTLPTraceExporter,
} = require('@opentelemetry/exporter-trace-otlp-proto');
const {
  OTLPLogExporter,
} = require('@opentelemetry/exporter-logs-otlp-proto');
const mysql = require('mysql2/promise');

// Test configuration
const TEST_RUN_ID = Date.now().toString();
const SERVICE_NAME = `otel-nodejs-test-${TEST_RUN_ID}`;

// Environment configuration
const config = {
  httpHost: process.env.HTTP_HOST || '127.0.0.1',
  httpPort: process.env.HTTP_PORT || '4000',
  mysqlHost: process.env.MYSQL_HOST || '127.0.0.1',
  mysqlPort: parseInt(process.env.MYSQL_PORT || '4002', 10),
  dbName: process.env.DB_NAME || 'otel_tests',
  username: process.env.GREPTIME_USERNAME || '',
  password: process.env.GREPTIME_PASSWORD || '',
};

// Get OTLP endpoint
const getOtlpEndpoint = () =>
  `http://${config.httpHost}:${config.httpPort}/v1/otlp`;

// Get OTLP headers
const getOtlpHeaders = () => {
  const headers = {
    'X-Greptime-DB-Name': config.dbName,
  };
  if (config.username || config.password) {
    const auth = Buffer.from(`${config.username}:${config.password}`).toString(
      'base64'
    );
    headers['Authorization'] = `Basic ${auth}`;
  }
  return headers;
};

// Providers
let meterProvider;
let tracerProvider;
let loggerProvider;
let dbConnection;

// Helper function for delay
const delay = (ms) => new Promise((resolve) => setTimeout(resolve, ms));

// Setup and teardown
beforeAll(async () => {
  console.log('\nInitializing OpenTelemetry:');
  console.log(`  OTLP endpoint: ${getOtlpEndpoint()}`);
  console.log(`  MySQL: ${config.mysqlHost}:${config.mysqlPort}/${config.dbName}`);
  console.log(`  Service name: ${SERVICE_NAME}`);

  // Create resource
  const resource = new Resource({
    [SEMRESATTRS_SERVICE_NAME]: SERVICE_NAME,
    [SEMRESATTRS_SERVICE_VERSION]: '1.0.0',
    'test.run.id': TEST_RUN_ID,
  });

  // Initialize Metrics
  const metricExporter = new OTLPMetricExporter({
    url: `${getOtlpEndpoint()}/v1/metrics`,
    headers: getOtlpHeaders(),
    timeoutMillis: 10000,
  });

  meterProvider = new MeterProvider({
    resource,
    readers: [
      new PeriodicExportingMetricReader({
        exporter: metricExporter,
        exportIntervalMillis: 1000,
      }),
    ],
  });
  metrics.setGlobalMeterProvider(meterProvider);

  // Initialize Traces
  // Note: Traces require x-greptime-pipeline-name header
  const traceHeaders = {
    ...getOtlpHeaders(),
    'x-greptime-pipeline-name': 'greptime_trace_v1',
  };

  const traceExporter = new OTLPTraceExporter({
    url: `${getOtlpEndpoint()}/v1/traces`,
    headers: traceHeaders,
    timeoutMillis: 10000,
  });

  tracerProvider = new BasicTracerProvider({ resource });
  tracerProvider.addSpanProcessor(new BatchSpanProcessor(traceExporter));
  tracerProvider.register();

  // Initialize Logs
  const logExporter = new OTLPLogExporter({
    url: `${getOtlpEndpoint()}/v1/logs`,
    headers: getOtlpHeaders(),
    timeoutMillis: 10000,
  });

  loggerProvider = new LoggerProvider({ resource });
  loggerProvider.addLogRecordProcessor(new BatchLogRecordProcessor(logExporter));
  logs.setGlobalLoggerProvider(loggerProvider);

  // Initialize database connection
  dbConnection = await mysql.createConnection({
    host: config.mysqlHost,
    port: config.mysqlPort,
    database: config.dbName,
    user: config.username || undefined,
    password: config.password || undefined,
  });

  console.log('OpenTelemetry initialized successfully');
}, 30000);

afterAll(async () => {
  console.log('\nShutting down OpenTelemetry...');

  if (meterProvider) {
    await meterProvider.shutdown();
  }
  if (tracerProvider) {
    await tracerProvider.shutdown();
  }
  if (loggerProvider) {
    await loggerProvider.shutdown();
  }
  if (dbConnection) {
    await dbConnection.end();
  }
}, 30000);

// ==================== METRICS TESTS ====================

describe('Metrics Tests', () => {
  test('Counter metric', async () => {
    console.log('\nTesting Metrics Counter...');

    const meter = metrics.getMeter('test-meter');
    const counterName = `otel_nodejs_counter_${TEST_RUN_ID}`;

    // Create counter
    const counter = meter.createCounter(counterName, {
      description: 'Test counter for OTLP integration',
      unit: '1',
    });

    // Record values
    counter.add(10, { env: 'test' });
    counter.add(25, { env: 'test' });
    counter.add(15, { env: 'prod' });

    // Force flush and wait
    await meterProvider.forceFlush();
    await delay(3000);

    // Verify via SQL
    // Note: GreptimeDB adds _total suffix for counters (Prometheus convention)
    const tableName = `${counterName}_total`;
    try {
      const [rows] = await dbConnection.execute(
        `SELECT * FROM ${tableName} WHERE env = 'test' ORDER BY greptime_timestamp DESC LIMIT 1`
      );
      console.log(
        `Counter verification: ${rows.length > 0 ? 'PASSED' : 'table may have different structure'}`
      );
    } catch (e) {
      console.log(`Query note: ${e.message}`);
    }

    console.log(`Counter test completed for table: ${tableName}`);
  }, 30000);

  test('Gauge metric', async () => {
    console.log('\nTesting Metrics Gauge...');

    const meter = metrics.getMeter('test-meter');
    const gaugeName = `otel_nodejs_gauge_${TEST_RUN_ID}`;

    // Create gauge (using UpDownCounter as gauge equivalent)
    const gauge = meter.createUpDownCounter(gaugeName, {
      description: 'Test gauge for OTLP integration',
      unit: 'bytes',
    });

    // Record values
    gauge.add(1024, { host: 'server1' });
    gauge.add(2048, { host: 'server2' });

    // Force flush and wait
    await meterProvider.forceFlush();
    await delay(3000);

    // Note: GreptimeDB adds _<unit> suffix for gauges (Prometheus convention)
    const tableName = `${gaugeName}_bytes`;
    console.log(`Gauge test completed for table: ${tableName}`);
  }, 30000);

  test('Histogram metric', async () => {
    console.log('\nTesting Metrics Histogram...');

    const meter = metrics.getMeter('test-meter');
    const histogramName = `otel_nodejs_histogram_${TEST_RUN_ID}`;

    // Create histogram
    const histogram = meter.createHistogram(histogramName, {
      description: 'Test histogram for OTLP integration',
      unit: 'ms',
    });

    // Record values
    histogram.record(15.5, { endpoint: '/api/users' });
    histogram.record(45.2, { endpoint: '/api/users' });
    histogram.record(120.8, { endpoint: '/api/orders' });
    histogram.record(5.1, { endpoint: '/health' });

    // Force flush and wait
    await meterProvider.forceFlush();
    await delay(3000);

    // Note: GreptimeDB converts "ms" to "milliseconds" and creates _bucket/_count/_sum tables
    const tableName = `${histogramName}_milliseconds_count`;
    console.log(`Histogram test completed for table: ${tableName}`);
  }, 30000);
});

// ==================== TRACES TESTS ====================

describe('Traces Tests', () => {
  test('Trace spans', async () => {
    console.log('\nTesting Traces...');

    const tracer = trace.getTracer('test-tracer');

    // Create parent span
    const parentSpan = tracer.startSpan(`parent-operation-${TEST_RUN_ID}`, {
      kind: SpanKind.SERVER,
      attributes: {
        'http.method': 'GET',
        'http.url': '/api/test',
        'test.run.id': TEST_RUN_ID,
      },
    });

    // Simulate work
    await delay(50);

    // Create child span
    const ctx = trace.setSpan(context.active(), parentSpan);
    const childSpan = tracer.startSpan(
      `child-operation-${TEST_RUN_ID}`,
      {
        kind: SpanKind.INTERNAL,
        attributes: {
          'db.system': 'greptimedb',
          'db.operation': 'query',
        },
      },
      ctx
    );

    await delay(30);
    childSpan.setStatus({ code: SpanStatusCode.OK });
    childSpan.end();

    parentSpan.setStatus({ code: SpanStatusCode.OK });
    parentSpan.end();

    // Force flush and wait
    await tracerProvider.forceFlush();
    await delay(3000);

    // Verify via SQL
    // Note: service_name is a direct column in the table
    try {
      const [rows] = await dbConnection.execute(
        `SELECT * FROM opentelemetry_traces WHERE service_name = '${SERVICE_NAME}' ORDER BY timestamp DESC LIMIT 2`
      );
      console.log(`Found ${rows.length} trace spans in opentelemetry_traces`);
    } catch (e) {
      console.log(`Query note: ${e.message}`);
    }

    console.log('Traces test completed');
  }, 30000);
});

// ==================== LOGS TESTS ====================

describe('Logs Tests', () => {
  test('Log records', async () => {
    console.log('\nTesting Logs...');

    const logger = logs.getLogger('test-logger');

    // Emit log records
    logger.emit({
      severityNumber: SeverityNumber.INFO,
      severityText: 'INFO',
      body: `Test info log message - ${TEST_RUN_ID}`,
      attributes: {
        'test.run.id': TEST_RUN_ID,
        'log.type': 'test',
      },
    });

    logger.emit({
      severityNumber: SeverityNumber.WARN,
      severityText: 'WARN',
      body: `Test warning log message - ${TEST_RUN_ID}`,
      attributes: {
        'test.run.id': TEST_RUN_ID,
        'log.type': 'test',
      },
    });

    logger.emit({
      severityNumber: SeverityNumber.ERROR,
      severityText: 'ERROR',
      body: `Test error log message - ${TEST_RUN_ID}`,
      attributes: {
        'test.run.id': TEST_RUN_ID,
        'log.type': 'test',
        'error.code': 'TEST_ERROR',
      },
    });

    // Force flush and wait
    await loggerProvider.forceFlush();
    await delay(3000);

    // Verify via SQL
    // Note: service_name is stored in resource_attributes, use body with TEST_RUN_ID to query
    try {
      const [rows] = await dbConnection.execute(
        `SELECT * FROM opentelemetry_logs WHERE body LIKE '%${TEST_RUN_ID}%' ORDER BY timestamp DESC LIMIT 3`
      );
      console.log(`Found ${rows.length} log records in opentelemetry_logs`);
    } catch (e) {
      console.log(`Query note: ${e.message}`);
    }

    console.log('Logs test completed');
  }, 30000);
});

// ==================== COMPREHENSIVE TEST ====================

describe('Comprehensive Tests', () => {
  test('All signals together', async () => {
    console.log('\nTesting all signals together...');

    const meter = metrics.getMeter('comprehensive-test');
    const tracer = trace.getTracer('comprehensive-test');
    const logger = logs.getLogger('comprehensive-test');

    // Create a trace span
    const span = tracer.startSpan(`comprehensive-operation-${TEST_RUN_ID}`, {
      kind: SpanKind.SERVER,
    });

    // Record metrics within the span
    const counterName = `otel_nodejs_comprehensive_${TEST_RUN_ID}`;
    const counter = meter.createCounter(counterName);
    counter.add(1, { operation: 'comprehensive' });

    // Emit a log within the span
    logger.emit({
      severityNumber: SeverityNumber.INFO,
      severityText: 'INFO',
      body: `Comprehensive test log - ${TEST_RUN_ID}`,
      attributes: { 'test.type': 'comprehensive' },
    });

    // Simulate work
    await delay(100);

    span.setStatus({ code: SpanStatusCode.OK });
    span.end();

    // Force flush all providers
    await meterProvider.forceFlush();
    await tracerProvider.forceFlush();
    await loggerProvider.forceFlush();
    await delay(3000);

    console.log('Comprehensive test completed successfully');
  }, 30000);
});

File: go-tests/otel_test.go
// Copyright 2023 Greptime Team
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package tests

import (
	"context"
	"database/sql"
	"encoding/base64"
	"fmt"
	"os"
	"sync"
	"testing"
	"time"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/codes"
	"go.opentelemetry.io/otel/exporters/otlp/otlplog/otlploghttp"
	"go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetrichttp"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp"
	"go.opentelemetry.io/otel/log"
	"go.opentelemetry.io/otel/metric"
	sdklog "go.opentelemetry.io/otel/sdk/log"
	sdkmetric "go.opentelemetry.io/otel/sdk/metric"
	"go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	semconv "go.opentelemetry.io/otel/semconv/v1.26.0"
	"go.opentelemetry.io/otel/trace"
)

var (
	otelTestRunID   = fmt.Sprintf("%d", time.Now().UnixMilli())
	otelServiceName = "otel-go-test-" + otelTestRunID

	// Providers (lazily initialized)
	otelMeterProvider  *sdkmetric.MeterProvider
	otelTracerProvider *sdktrace.TracerProvider
	otelLoggerProvider *sdklog.LoggerProvider
	otelDB             *sql.DB
	otelInitOnce       sync.Once
	otelInitErr        error
	otelInitialized    bool
)

// setupOtel initializes OpenTelemetry providers (called once via sync.Once)
func setupOtel(t *testing.T) {
	otelInitOnce.Do(func() {
		ctx := context.Background()

		httpHost := getEnv("HTTP_HOST", "127.0.0.1")
		httpPort := getEnv("HTTP_PORT", "4000")
		mysqlHost := getEnv("MYSQL_HOST", "127.0.0.1")
		mysqlPort := getEnv("MYSQL_PORT", "4002")
		dbName := getEnv("DB_NAME", "go_tests")
		username := getEnv("GREPTIME_USERNAME", "")
		password := getEnv("GREPTIME_PASSWORD", "")

		endpoint := fmt.Sprintf("%s:%s", httpHost, httpPort)

		// Build headers
		headers := map[string]string{
			"X-Greptime-DB-Name": dbName,
		}
		if username != "" || password != "" {
			auth := base64.StdEncoding.EncodeToString([]byte(username + ":" + password))
			headers["Authorization"] = "Basic " + auth
		}

		// Create resource
		res, err := resource.New(ctx,
			resource.WithAttributes(
				semconv.ServiceName(otelServiceName),
				semconv.ServiceVersion("1.0.0"),
				attribute.String("test.run.id", otelTestRunID),
			),
		)
		if err != nil {
			otelInitErr = fmt.Errorf("failed to create resource: %w", err)
			return
		}

		// Initialize Metrics
		metricExporter, err := otlpmetrichttp.New(ctx,
			otlpmetrichttp.WithEndpoint(endpoint),
			otlpmetrichttp.WithURLPath("/v1/otlp/v1/metrics"),
			otlpmetrichttp.WithHeaders(headers),
			otlpmetrichttp.WithInsecure(),
		)
		if err != nil {
			otelInitErr = fmt.Errorf("failed to create metric exporter: %w", err)
			return
		}

		otelMeterProvider = sdkmetric.NewMeterProvider(
			sdkmetric.WithResource(res),
			sdkmetric.WithReader(sdkmetric.NewPeriodicReader(metricExporter,
				sdkmetric.WithInterval(time.Second),
			)),
		)
		otel.SetMeterProvider(otelMeterProvider)

		// Initialize Traces
		traceHeaders := make(map[string]string)
		for k, v := range headers {
			traceHeaders[k] = v
		}
		traceHeaders["x-greptime-pipeline-name"] = "greptime_trace_v1"

		traceExporter, err := otlptracehttp.New(ctx,
			otlptracehttp.WithEndpoint(endpoint),
			otlptracehttp.WithURLPath("/v1/otlp/v1/traces"),
			otlptracehttp.WithHeaders(traceHeaders),
			otlptracehttp.WithInsecure(),
		)
		if err != nil {
			otelInitErr = fmt.Errorf("failed to create trace exporter: %w", err)
			return
		}

		otelTracerProvider = sdktrace.NewTracerProvider(
			sdktrace.WithResource(res),
			sdktrace.WithBatcher(traceExporter),
		)
		otel.SetTracerProvider(otelTracerProvider)

		// Initialize Logs
		logExporter, err := otlploghttp.New(ctx,
			otlploghttp.WithEndpoint(endpoint),
			otlploghttp.WithURLPath("/v1/otlp/v1/logs"),
			otlploghttp.WithHeaders(headers),
			otlploghttp.WithInsecure(),
		)
		if err != nil {
			otelInitErr = fmt.Errorf("failed to create log exporter: %w", err)
			return
		}

		otelLoggerProvider = sdklog.NewLoggerProvider(
			sdklog.WithResource(res),
			sdklog.WithProcessor(sdklog.NewBatchProcessor(logExporter)),
		)

		// Initialize DB connection
		dsn := fmt.Sprintf("%s:%s@tcp(%s:%s)/%s", username, password, mysqlHost, mysqlPort, dbName)
		otelDB, err = sql.Open("mysql", dsn)
		if err != nil {
			otelInitErr = fmt.Errorf("failed to open database: %w", err)
			return
		}
		if err = otelDB.Ping(); err != nil {
			otelInitErr = fmt.Errorf("failed to ping database: %w", err)
			return
		}

		t.Logf("OpenTelemetry initialized: endpoint=%s, service=%s", endpoint, otelServiceName)
		otelInitialized = true
	})

	if otelInitErr != nil {
		t.Fatalf("Failed to initialize OpenTelemetry: %v", otelInitErr)
	}
}

// cleanupOtel shuts down all OpenTelemetry providers (called from TestMain)
func cleanupOtel() {
	if !otelInitialized {
		return
	}
	ctx := context.Background()
	if otelMeterProvider != nil {
		otelMeterProvider.Shutdown(ctx)
	}
	if otelTracerProvider != nil {
		otelTracerProvider.Shutdown(ctx)
	}
	if otelLoggerProvider != nil {
		otelLoggerProvider.Shutdown(ctx)
	}
	if otelDB != nil {
		otelDB.Close()
	}
}

// TestMain handles package-level setup and teardown
func TestMain(m *testing.M) {
	code := m.Run()
	cleanupOtel()
	os.Exit(code)
}

// verifyTableHasData queries the table and asserts it has data
// Returns the count of rows found
func verifyTableHasData(t *testing.T, tableName string, whereClause string, expectedMin int) int {
	t.Helper()

	if otelDB == nil {
		t.Logf("Skipping verification: database connection not available")
		return 0
	}

	query := fmt.Sprintf("SELECT COUNT(*) FROM %s", tableName)
	if whereClause != "" {
		query += " WHERE " + whereClause
	}

	var count int
	err := otelDB.QueryRowContext(context.Background(), query).Scan(&count)
	if err != nil {
		t.Fatalf("Failed to query table %s: %v\n  Query: %s", tableName, err, query)
	}

	if count < expectedMin {
		t.Errorf("Table %s has %d rows, expected at least %d\n  Query: %s", tableName, count, expectedMin, query)
	} else {
		t.Logf("Verified: table %s has %d rows (expected >= %d)", tableName, count, expectedMin)
	}

	return count
}

// ==================== METRICS TESTS ====================

func TestOtelMetricsCounter(t *testing.T) {
	setupOtel(t)

	ctx := context.Background()
	meter := otel.Meter("test-meter")

	counterName := "otel_go_counter_" + otelTestRunID
	counter, err := meter.Int64Counter(counterName,
		metric.WithDescription("Test counter for OTLP integration"),
		metric.WithUnit("1"),
	)
	if err != nil {
		t.Fatalf("Failed to create counter: %v", err)
	}

	t.Log("Recording counter values: 10 (env=test), 25 (env=test), 15 (env=prod)")
	counter.Add(ctx, 10, metric.WithAttributes(attribute.String("env", "test")))
	counter.Add(ctx, 25, metric.WithAttributes(attribute.String("env", "test")))
	counter.Add(ctx, 15, metric.WithAttributes(attribute.String("env", "prod")))

	if err := otelMeterProvider.ForceFlush(ctx); err != nil {
		t.Logf("Warning: Force flush failed: %v", err)
	}
	time.Sleep(3 * time.Second)

	// Verify: counter creates table with _total suffix (Prometheus convention)
	tableName := counterName + "_total"
	verifyTableHasData(t, tableName, "", 1)
}

func TestOtelMetricsGauge(t *testing.T) {
	setupOtel(t)

	ctx := context.Background()
	meter := otel.Meter("test-meter")

	gaugeName := "otel_go_gauge_" + otelTestRunID
	gauge, err := meter.Int64Gauge(gaugeName,
		metric.WithDescription("Test gauge for OTLP integration"),
		metric.WithUnit("bytes"),
	)
	if err != nil {
		t.Fatalf("Failed to create gauge: %v", err)
	}

	t.Log("Recording gauge values: 1024 (host=server1), 2048 (host=server2)")
	gauge.Record(ctx, 1024, metric.WithAttributes(attribute.String("host", "server1")))
	gauge.Record(ctx, 2048, metric.WithAttributes(attribute.String("host", "server2")))

	if err := otelMeterProvider.ForceFlush(ctx); err != nil {
		t.Logf("Warning: Force flush failed: %v", err)
	}
	time.Sleep(3 * time.Second)

	// Verify: gauge creates table with _<unit> suffix (Prometheus convention)
	tableName := gaugeName + "_bytes"
	verifyTableHasData(t, tableName, "", 1)
}

func TestOtelMetricsHistogram(t *testing.T) {
	setupOtel(t)

	ctx := context.Background()
	meter := otel.Meter("test-meter")

	histogramName := "otel_go_histogram_" + otelTestRunID
	histogram, err := meter.Float64Histogram(histogramName,
		metric.WithDescription("Test histogram for OTLP integration"),
		metric.WithUnit("ms"),
	)
	if err != nil {
		t.Fatalf("Failed to create histogram: %v", err)
	}

	t.Log("Recording histogram values: 15.5ms, 45.2ms, 120.8ms, 5.1ms")
	histogram.Record(ctx, 15.5, metric.WithAttributes(attribute.String("endpoint", "/api/users")))
	histogram.Record(ctx, 45.2, metric.WithAttributes(attribute.String("endpoint", "/api/users")))
	histogram.Record(ctx, 120.8, metric.WithAttributes(attribute.String("endpoint", "/api/orders")))
	histogram.Record(ctx, 5.1, metric.WithAttributes(attribute.String("endpoint", "/health")))

	if err := otelMeterProvider.ForceFlush(ctx); err != nil {
		t.Logf("Warning: Force flush failed: %v", err)
	}
	time.Sleep(3 * time.Second)

	// Verify: histogram creates _count table (also _sum, _bucket)
	// Note: "ms" unit is converted to "milliseconds" (Prometheus convention)
	tableName := histogramName + "_milliseconds_count"
	verifyTableHasData(t, tableName, "", 1)
}

// ==================== TRACES TESTS ====================

func TestOtelTraces(t *testing.T) {
	setupOtel(t)

	ctx := context.Background()
	tracer := otel.Tracer("test-tracer")

	t.Logf("Creating trace spans for service: %s", otelServiceName)
	ctx, parentSpan := tracer.Start(ctx, "parent-operation-"+otelTestRunID,
		trace.WithSpanKind(trace.SpanKindServer),
		trace.WithAttributes(
			attribute.String("http.method", "GET"),
			attribute.String("http.url", "/api/test"),
			attribute.String("test.run.id", otelTestRunID),
		),
	)

	time.Sleep(50 * time.Millisecond)

	_, childSpan := tracer.Start(ctx, "child-operation-"+otelTestRunID,
		trace.WithSpanKind(trace.SpanKindInternal),
		trace.WithAttributes(
			attribute.String("db.system", "greptimedb"),
			attribute.String("db.operation", "query"),
		),
	)
	time.Sleep(30 * time.Millisecond)
	childSpan.SetStatus(codes.Ok, "")
	childSpan.End()

	parentSpan.SetStatus(codes.Ok, "")
	parentSpan.End()
	t.Log("Trace spans created: parent-operation, child-operation")

	if err := otelTracerProvider.ForceFlush(ctx); err != nil {
		t.Logf("Warning: Force flush failed: %v", err)
	}
	time.Sleep(3 * time.Second)

	// Verify: traces are stored in opentelemetry_traces table
	whereClause := fmt.Sprintf("service_name = '%s'", otelServiceName)
	verifyTableHasData(t, "opentelemetry_traces", whereClause, 2)
}

// ==================== LOGS TESTS ====================

func TestOtelLogs(t *testing.T) {
	setupOtel(t)

	ctx := context.Background()
	logger := otelLoggerProvider.Logger("test-logger")

	t.Logf("Emitting 3 log records with test.run.id=%s", otelTestRunID)

	var record log.Record
	record.SetSeverity(log.SeverityInfo)
	record.SetBody(log.StringValue("Test info log message - " + otelTestRunID))
	record.AddAttributes(
		log.String("test.run.id", otelTestRunID),
		log.String("log.type", "test"),
	)
	logger.Emit(ctx, record)

	record = log.Record{}
	record.SetSeverity(log.SeverityWarn)
	record.SetBody(log.StringValue("Test warning log message - " + otelTestRunID))
	record.AddAttributes(
		log.String("test.run.id", otelTestRunID),
		log.String("log.type", "test"),
	)
	logger.Emit(ctx, record)

	record = log.Record{}
	record.SetSeverity(log.SeverityError)
	record.SetBody(log.StringValue("Test error log message - " + otelTestRunID))
	record.AddAttributes(
		log.String("test.run.id", otelTestRunID),
		log.String("log.type", "test"),
		log.String("error.code", "TEST_ERROR"),
	)
	logger.Emit(ctx, record)
	t.Log("Log records emitted: INFO, WARN, ERROR")

	if err := otelLoggerProvider.ForceFlush(ctx); err != nil {
		t.Logf("Warning: Force flush failed: %v", err)
	}
	time.Sleep(3 * time.Second)

	// Verify: logs are stored in opentelemetry_logs table
	whereClause := fmt.Sprintf("body LIKE '%%%s%%'", otelTestRunID)
	verifyTableHasData(t, "opentelemetry_logs", whereClause, 3)
}

// ==================== COMPREHENSIVE TEST ====================

func TestOtelAllSignalsTogether(t *testing.T) {
	setupOtel(t)

	ctx := context.Background()
	meter := otel.Meter("comprehensive-test")
	tracer := otel.Tracer("comprehensive-test")
	logger := otelLoggerProvider.Logger("comprehensive-test")

	t.Log("Testing all signals together: metrics + traces + logs")

	ctx, span := tracer.Start(ctx, "comprehensive-operation-"+otelTestRunID,
		trace.WithSpanKind(trace.SpanKindServer),
	)

	counterName := "otel_go_comprehensive_" + otelTestRunID
	counter, _ := meter.Int64Counter(counterName)
	counter.Add(ctx, 1, metric.WithAttributes(attribute.String("operation", "comprehensive")))

	var record log.Record
	record.SetSeverity(log.SeverityInfo)
	record.SetBody(log.StringValue("Comprehensive test log - " + otelTestRunID))
	record.AddAttributes(log.String("test.type", "comprehensive"))
	logger.Emit(ctx, record)

	time.Sleep(100 * time.Millisecond)

	span.SetStatus(codes.Ok, "")
	span.End()

	otelMeterProvider.ForceFlush(ctx)
	otelTracerProvider.ForceFlush(ctx)
	otelLoggerProvider.ForceFlush(ctx)
	time.Sleep(3 * time.Second)

	// Verify all signals
	t.Log("Verifying all signals were exported...")
	tableName := counterName + "_total"
	verifyTableHasData(t, tableName, "", 1)
}
